### ISSUE TITLE
Add MarkupExtension , x:Static, x:Type, RelativeSource FindAncestor

### ISSUE THREAD

**Issue by andyb1979** [Microsoft/xaml-standard#27]
Without MarkupExtension, UWP is really really hard to develop controls for. See [this thread](https://github.com/Microsoft/Windows-universal-samples/issues/338) for background.

I am the owner of [SciChart ](https://www.scichart.com) - High Performance Realtime Charts. We make successful, big-data charts on WPF, iOS and Android platforms, and are close to releasing Xamarin.iOS and Xamarin.Android support. 

We don't support UWP. The reason why is it is virtually impossible to port our existing XAML which depends heavily on the x:Static extension, RelativeSource FindAncestor, and various custom MarkupExtensions which enable our default themeing with binding fallbacks. 

Example of code we want to replicate in UWP/Xaml Standard:

`    
    <Style x:Key="AnnotationBaseStyle" TargetType="a:AnnotationBase">
        <Style.Setters>
            <Setter Property="IsTabStop" Value="False" />
            <Setter Property="Foreground" Value="{me:ThemeBinding TickTextBrush}" />
            <Setter Property="FontSize" Value="12" />
        </Style.Setters>
    </Style>`

... where ThemeBinding is a class which inherits MarkupExtension. 

`<s3D:SciChart3DSurface x:Name="SciChart"
                               Grid.Column="1"
                               BorderThickness="0"
                               RenderableSeries="{extensions:SeriesBinding RenderableSeries}"
                               WorldDimensions="200,100,200">

            <s3D:SciChart3DSurface.Camera>
                <s3D:Camera3D ZoomToFitOnAttach="True" />
            </s3D:SciChart3DSurface.Camera>

            <s3D:SciChart3DSurface.Resources>
                <Style x:Key="SurfaceMeshStyle" TargetType="renderableSeries:SurfaceMeshRenderableSeries3DForMvvm">
                    <Setter Property="MeshColorPalette" Value="{StaticResource HeightColorMap}" />
                    <Setter Property="Stroke" Value="#77228B22" />
                </Style>

                <Style x:Key="WaterfallStyle" TargetType="renderableSeries:WaterfallRenderableSeries3DForMvvm">
                    <Setter Property="YColorMapping" Value="{StaticResource GradientColorPalette}" />
                    <Setter Property="ZColorMapping" Value="{StaticResource GradientColorPalette}" />
                </Style>
            </s3D:SciChart3DSurface.Resources>

            <s3D:SciChart3DSurface.XAxis>
                <s3D:NumericAxis3D AutoRange="Always" GrowBy="0.1,0.1" />
            </s3D:SciChart3DSurface.XAxis>

            <s3D:SciChart3DSurface.YAxis>
                <s3D:NumericAxis3D AutoRange="Always" GrowBy="0.1,0.1" />
            </s3D:SciChart3DSurface.YAxis>

            <s3D:SciChart3DSurface.ZAxis>
                <s3D:NumericAxis3D AutoRange="Always" GrowBy="0.1,0.1" />
            </s3D:SciChart3DSurface.ZAxis>

        </s3D:SciChart3DSurface>`

... where SeriesBinding is a markup extension that enables a complex binding in our library.

**Why is this so critical to us?**

- Without RelativeSource FindAncestor you have to resort to ugly hacks to bind to parts of the Visual Tree above the current element. [See my answer on StackOVerflow for guidance](http://stackoverflow.com/a/43399114/303612). 

 - Without x:Static it is impossible to bind to a static class, e.g. a global property ThemeManager.EnableDropShadows cannot be bound and get updates from it as the property changes. 

 - Without MarkupExtension it is impossible to code our own replacement for x:Static or any other custom markup other than Binding, ThemeResource in UWP. 

This means our only options are attached properties and custom parsing of complex strings. A nasty way to write xaml and we dont want to do it!

**Comment by h82258652** [comment]
I hope they can bring back x:Static and DynamicResource.
But I dont think ```RelativeSource.FindAncestor``` is necessary. In most case you can use binding element name to do this, and ```FindAncestor``` performance is not good.

**Comment by stefanolson** [comment]
I have been on about this for years.. I feel like no one in Microsoft actually listens to people who have invested in WPF. I think find ancestor binding is absolutely critical. I've got a lot of code where find ancestor is the only way to make it work and therefore I can't port things to UWP.  I'd also like to see multi-binding there are scenarios whether it is the best way to do things although if x:Bind were supported in WPF that may be less of an issue.

**Comment by andyb1979** [comment]
Yes RelativeSource FindAncestor is important.

Imagine you are making a custom control which has many templates parts. One
of those templates needs to get a value from a parent of a parent.

I can't give you any concrete examples save to say in the SciChart control
templates (generic.xaml) we use RelativeSource FindAncestor,
MarkupExtension and x:Static *all the time*. The lack of these prevent us
for developing for UWP.

**Comment by stefanolson** [comment]
I definitely have examples where I've had to create thousands of lines of C# code to work around lack of find ancestor.  It just makes things so much more complicated and messy.

**Comment by andyb1979** [comment]
@stefanolson  have you seen my workaround for [RelativeSource FindAncestor on UWP](http://stackoverflow.com/a/43399114/303612)? It works but it's ugly. 

That one was easy to code around but MarkupExtension, x:Static is not. I am considering to make attached roperties where I pass in a completely un-type-safe string and parse it e.g. 

Imagine I want to bind the Forground property statically to SomeType.Instance.SomeProperty. 

I can theoretically achieve it in UWP with the most egrereous hack. Creating a class with an attached property, that accepts a string which we parse to get the Target, the Source (static class) and the Path on the static path: 

```xml
<!-- Workaround for x:Static in UWP ... -->
<Setter Property="MyStaticHelper" Value="TargetProperty=Foreground, StaticClass=SomeType, StaticPath=Instance.SomeProperty">
</Setter>
```

You can imagine the sorts of issues here. We have to write code like this and hope that it works at runtime as the strings are not type safe. 

Of course, with MarkupExtension in UWP we can write our own Static, [which I have done for Silverlight in the past](https://www.codeproject.com/Articles/305932/Static-and-Type-markup-extensions-for-Silverlight).

**Comment by Mike-EEE** [comment]
FWIW, I've made a request on this UserVoice issue for the issue here to be the official GitHub equivalent for discussion and collaboration:

https://wpdev.uservoice.com/forums/110705-universal-windows-platform/suggestions/7232264-add-markup-extensions-to-and-improve-winrt-xaml

**Comment by andyb1979** [comment]
Hey Mike,

Yep myself and the SciChart team have already voted for it! Thanks!

-Andrew

**Comment by tomzorz** [comment]
<off>
@andyb1979 it's not impossible to bind to a static without markup extensions in UWP: you can pull it off with some CustomResource magic
</off>

**Comment by andyb1979** [comment]
Yes, or an attached property which accepts a complex string including Type,
Membername / path. However should we have to? These sort of features which
we have come to rely on for over a decade should be built in.

**Comment by tomzorz** [comment]
@andyb1979 Don't get me wrong, I wholeheartedly agree that we need MarkupExtensions. Just wanted to point it out that it was possible if you really wanted ;)

**Comment by birbilis** [comment]
MarkupExtensions were mentioned as being looked into for inclusion at XAML Standard at Build panel

**Comment by Mike-EEE** [comment]
> MarkupExtensions were mentioned as being looked into for inclusion at XAML Standard at Build panel

Not so fast.  Have you read the actual UserVoice comments lately?  They are not even set on using IServiceProvider or any of the core/abundant/valuable services that are available in WPF and pretty much every flavor of Xaml out there.  That hardly classifies as a MarkupExtension, I'm afraid.

So when a panel states that MEs are being included in the standard, what is the definition of an ME, exactly?  This is a slightly important detail to get squared away.

(Also, thank you for your support, @andyb1979!)

**Comment by birbilis** [comment]
They probably care about the syntax at the XAML level, but may want to introduce some other implementation at the codebehind.
